feat(ldap.dn): Add support for different formats in `ldap.dn2str()` via flags

spaceone opened the pull request:
In C `dn2str()` supports `flags` which works by providing `LDAP_DN_FORMAT_UFN`, `LDAP_DN_FORMAT_AD_CANONICAL`, ….
These symbols do exist in Python, but could not be used ultimately because the Python counterpart was pure Python and did not pass to `dn2str(3)`.

Fix #257

spaceone committed:
test(ldap.dn): Add test cases for ldap.dn.dn2str() with different format flags

spaceone committed:
feat(ldap.dn): Add support for different formats in `ldap.dn.dn2str()` via flags

In C `dn2str()` supports `flags` which works by providing one of `LDAP_DN_FORMAT_UFN`, `LDAP_DN_FORMAT_AD_CANONICAL`, `LDAP_DN_FORMAT_DCE`, `LDAP_DN_FORMAT_LDAPV3`.
These symbols do exist in Python, but could not be used ultimately because the Python counterpart was pure Python and did not pass to `dn2str(3)`.

Fix #257

droideck commented:
@spaceone hi!
The PR says 'WIP', - is it still true?

spaceone commented:
> @spaceone hi! The PR says 'WIP', - is it still true?

@droideck Well, I would say it needs code clenaup but the functionality is done. But before I try to cleanup I would like to get feedback.

droideck commented:
Overall, I think it'll be nice to have the feature.
@mistotebe what do you think?

mistotebe commented:
I agree, getting better DN handling would be nice.

spaceone commented on Tests/t_ldap_dn.py:
```diff
@@ -158,15 +158,42 @@ def test_dn2str(self):
                 [('dc', 'com', 1)]
             ]),
-            'uid=test\\, 42,ou=Testing,dc=example,dc=com'
+            'uid=test\\2C 42,ou=Testing,dc=example,dc=com'
```
API change: I don't find a way to prevent this

mistotebe replied:
Even though libldap does some DN normalisation, we can't guarantee to our users that the output of `dn2str` is going to be (character to character) identical. Maybe we should make it explicit that the DN they get from it is always going to be "the same one" but the string representation might evolve without notice? Certainly something that makes them think twice before using it as a dict key, etc.

spaceone commented on Tests/t_ldap_dn.py:
```diff
@@ -158,15 +158,42 @@ def test_dn2str(self):
                 [('dc', 'com', 1)]
             ]),
-            'cn=äöüÄÖÜß,dc=example,dc=com'
+            r'cn=\C3\A4\C3\B6\C3\BC\C3\84\C3\96\C3\9C\C3\9F,dc=example,dc=com'
```
API change: I don't find a way to prevent this

mistotebe replied:
This is definitely intentional and the old value incorrect, we're telling libldap to treat the value as non-printable (`LDAP_AVA_NONPRINTABLE == 0x04`) and escape accordingly.

spaceone replied:
ah yes, I overlooked that!

spaceone replied:
Same vor `LDAP_AVA_STRING`: 
```python
>>> ldap.dn.dn2str([[('cn', 'äöüÄÖÜß', ldap.AVA_BINARY)],[('dc', 'example', 1)],[('dc', 'com', 1)]])
'cn=#C3A4C3B6C3BCC384C396C39CC39F,dc=example,dc=com'
>>> ldap.dn.dn2str([[('cn', 'äöüÄÖÜß', ldap.AVA_NULL)],[('dc', 'example', 1)],[('dc', 'com', 1)]])
'cn=\\C3\\A4\\C3\\B6\\C3\\BC\\C3\\84\\C3\\96\\C3\\9C\\C3\\9F,dc=example,dc=com'
>>> ldap.dn.dn2str([[('cn', 'äöüÄÖÜß', ldap.AVA_STRING)],[('dc', 'example', 1)],[('dc', 'com', 1)]])
'cn=\\C3\\A4\\C3\\B6\\C3\\BC\\C3\\84\\C3\\96\\C3\\9C\\C3\\9F,dc=example,dc=com'
>>> ldap.dn.dn2str([[('cn', 'äöüÄÖÜß', ldap.AVA_NONPRINTABLE)],[('dc', 'example', 1)],[('dc', 'com', 1)]])
'cn=\\C3\\A4\\C3\\B6\\C3\\BC\\C3\\84\\C3\\96\\C3\\9C\\C3\\9F,dc=example,dc=com'
```

spaceone replied:
I added all of them as test cases.

spaceone commented on Lib/ldap/dn.py:
```diff
@@ -48,18 +48,19 @@ def str2dn(dn,flags=0):
-      for atype,avalue,dummy in rdn])
-    for rdn in dn
-  ])
+  #if not dn:
+  #    return ''
```
required?

mistotebe replied:
libldap isn't too sure either, but allows it: https://git.openldap.org/openldap/openldap/-/blob/OPENLDAP_REL_ENG_2_6_3/libraries/libldap/getdn.c#L3021-3028

mistotebe replied:
Unless we *want* to make the distinction for `None` and make `dn2str(None) == None`, might be too much hassle though.

spaceone committed:
feat(ldap.dn): add ldap.dn.normalize()

spaceone commented:
Performance comparison:

C:
```
python3 -m timeit -s 'import ldap.dn' "ldap.dn.dn2str([[('uid', 'test, 42', 1)],[('ou', 'Testing', 1)],[('dc', 'example', 1)],[('dc', 'com', 1)]], ldap.DN_FORMAT_LDAPV3)"
200000 loops, best of 5: 1.41 usec per loop
``` 
Python:
```
python3 -m timeit -s 'import ldap.dn' "ldap.dn.dn2str([[('uid', 'test, 42', 1)],[('ou', 'Testing', 1)],[('dc', 'example', 1)],[('dc', 'com', 1)]], 0)"
50000 loops, best of 5: 4.1 usec per loop
```

mistotebe reviewed (dismissed):
LGTM, maybe be consistent about using raw strings (`r'cn=\C3...'` vs. `'cn=\\C3...'`)

spaceone committed:
style(Tests/t_ldap_dn): use raw strings instead of escape sequences

droideck commented on Modules/functions.c:
```diff
@@ -160,6 +160,222 @@ l_ldap_str2dn(PyObject *unused, PyObject *args)
+    if (dn == NULL)
+        return;
+
+    for (LDAPRDN * rdn = dn; *rdn != NULL; rdn++) {
+        for (LDAPAVA ** avap = *rdn; *avap != NULL; avap++) {
```
Minor style typo:
```suggestion
    for (LDAPRDN *rdn = dn; *rdn != NULL; rdn++) {
        for (LDAPAVA **avap = *rdn; *avap != NULL; avap++) {
```

droideck reviewed (dismissed):
Looks good to me!

Could you please add a couple of more tests?
For NUL handling, for huge DNs, and for invalid flags.

If not, no worries, we can improve it later.

spaceone commented:
> Looks good to me!
Thank you!

> Could you please add a couple of more tests?
> For NUL handling
I added every possible combination I can think of :-)

> for huge DNs
Okay, added multiple string lengthes.

> and for invalid flags.
this already existed.

droideck reviewed (dismissed):
Looks good to me!

Please, squash the commits if you want (or confirm if you want to have it as separate commits).

spaceone commented:
The should not be squashed, they follow [conventional commit specification](https://www.conventionalcommits.org/en/v1.0.0/).

spaceone commented:
@droideck When can I expect a  release containing this (backwards compatible) feature? Since month there are no commits again. And even after this is contained in a release, I have to wait a specific time, so that it appears in Debian, too. It would enhance our build process a lot if we could use official releases instead of building everything on our own.
